kernel/kernel-api/src/main/java/io/delta/kernel/internal/stats/FileSizeHistogram.java

```diff
@@ -0,0 +1,248 @@
+  }
+
+  /**
+   * Adds a file size to the histogram, incrementing the appropriate bin's count and total bytes.
+   *
+   * @param fileSize The size of the file in bytes
+   * @throws IllegalArgumentException if fileSize is negative
+   */
+  public void insert(long fileSize) {
+    checkArgument(fileSize >= 0, "File size must be non-negative, got %s", fileSize);
+    int index = getBinIndex(fileSize);
+    checkArgument(index >= 0,
+            "getBinIndex must return non-negative index for non-negative fileSize, got %s", index);
+    fileCounts[index]++;
+    totalBytes[index] += fileSize;
+  }
+
+  /**
+   * Removes a file size from the histogram, decrementing the appropriate bin's count and total
+   * bytes.
+   *
+   * @param fileSize The size of the file in bytes
+   * @throws IllegalArgumentException if fileSize is negative
+   */
+  public void remove(long fileSize) {
+    checkArgument(fileSize >= 0, "File size must be non-negative, got %s", fileSize);
+    int index = getBinIndex(fileSize);
+    checkArgument(index >= 0,
+            "getBinIndex must return non-negative index for non-negative fileSize, got %s", index);
+    checkArgument(
+            totalBytes[index] >= fileSize,
+            "Cannot remove %s bytes from bin %d which only has %s bytes",
+            fileSize,
+            index,
+            totalBytes[index]);
+    fileCounts[index]--;
+    totalBytes[index] -= fileSize;
+  }
+
+  /**
+   * Encode as a {@link Row} object with the schema {@link FileSizeHistogram#FULL_SCHEMA}.
+   *
+   * @return {@link Row} object with the schema {@link FileSizeHistogram#FULL_SCHEMA}
+   */
+  public Row toRow() {
+    Map<Integer, Object> value = new HashMap<>();
+    value.put(
+        FULL_SCHEMA.indexOf("sortedBinBoundaries"),
+        VectorUtils.buildArrayValue(sortedBinBoundaries, LongType.LONG));
+    value.put(
+        FULL_SCHEMA.indexOf("fileCounts"),
+        VectorUtils.buildArrayValue(
+            Arrays.stream(fileCounts).collect(Collectors.toList()), LongType.LONG));
+    value.put(
+        FULL_SCHEMA.indexOf("totalBytes"),
+        VectorUtils.buildArrayValue(
+            Arrays.stream(totalBytes).collect(Collectors.toList()), LongType.LONG));
+    return new GenericRow(FULL_SCHEMA, value);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (o == null || getClass() != o.getClass()) return false;
+
+    FileSizeHistogram that = (FileSizeHistogram) o;
+    return sortedBinBoundaries.equals(that.sortedBinBoundaries)
+        && Arrays.equals(fileCounts, that.fileCounts)
+        && Arrays.equals(totalBytes, that.totalBytes);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(
+        sortedBinBoundaries, Arrays.hashCode(fileCounts), Arrays.hashCode(totalBytes));
+  }
+
+  private int getBinIndex(long fileSize) {
+    int searchResult = Collections.binarySearch(sortedBinBoundaries, fileSize);
+    return searchResult >= 0 ? searchResult : -(searchResult + 1) - 1;
```

Comment:
Can you explain this please? `-(searchResult + 1) - 1`

Reply:
Added comments, mainly if not found Arrays.binarySearch will return ( -insert point - 1)